docs/sources/operations/autoscaling_queriers.md

```diff
@@ -0,0 +1,139 @@
+---
+title: Autoscaling Loki Queriers
+weight: 30
+---
+
+# Autoscaling Loki Queriers
+
+A microservices deployment of a Loki cluster that runs on Kubernetes will typically handle a
+workload that varies throughout the day.
+To make Loki easier to operate and optimize the cost of running Loki at scale,
+we have designed a set of resources to help you autoscale your Loki queriers.
+
+## Prerequisites
+
+You need to run Loki in Kubernetes as a set of microservices.
+
+We recommend using [KEDA (*Kubernetes-based Event Driven Autoscaler*)](https://keda.sh/) to configure autoscaling
+based on Prometheus metrics. Refer to the [KEDA documentation](https://keda.sh/docs/latest/deploy) to learn more
+about setting up KEDA in your Kubernetes cluster.
+
+## Scaling metric
+
+Queriers pull queries from the query scheduler queue and process them on the querier workers. Therefore, it makes sense to scale based on:
+
+- The scheduler queue size.
+- The queries running in the queriers.
+
+The Query Scheduler exposes the `cortex_query_scheduler_inflight_requests` metric.
+It tracks the number of queued queries plus the number of queries currently running in the querier workers.
+The following query is useful to scale queriers based on the inflight requests.
+
+```promql
+sum(
+  max_over_time(
+    cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="<Q>"}[<R>]
+  )
+)
+```
+
+We can use the quantile (Q) and the range (R) parameters to fine-tune the metric.
+The higher Q is, the more sensitive the metric is to short-lasting spikes.
+As R increases, we reduce the variation over time in the metric.
+A higher value for R helps avoid the autoscaler from modifying the number of replicas too frequently.
+
+In our experience, we have found that a Q of 0.75 and R of 2 minutes work well.
+Our recommendation is to adjust these values according to your workload.
+
+## Cluster capacity planning
+
+When configuring KEDA to scale your Loki queriers, we need to configure a few things:
+
+- The threshold for scaling up and down.
+- Scale down stabilization period.
+- The minimum and the maximum number of queriers.
+
+Querier workers process queries from the queue, each of our queriers can be configured to run several workers.
+Our recommendation is to aim using 75% of those workers to leave some workforce headroom for workload spikes.
+Therefore, if we configure our queriers to run 6 workers, we will set a threshold of `floor(0.75 * 6) = 4`.
+
+Our recommendation for the minimum number of queriers is running at least one queriers and look at the average
+number of inflight requests 75% of the time in the last seven days, targeting a 75% utilization of our queriers.
+So if we use 6 workers per querier, we will use the following query:
+
+```promql
+clamp_min(ceil(
+    avg(
+        avg_over_time(cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="0.75"}[7d])
+    ) / scalar(floor(vector(6 * 0.75)))
+), 1)
+```
```

Comment:
If I understand correctly, you want that 75% of the time, the average number of request in the system (`L` in queuing theory) is equal to 75% the number of workers per querier, hence, there would be no query in the queue (they would be addressed immediately).
By evaluating the ratio between the **current** `L` (thanks to the inflight metric) and our **target** `L = 0.75 * nb_worker` you know how much times we need to scale the **current** number of querier to reach our **target**.

Side note, this is only meaningful if `lambda` (the request arrival rate) converge over time which is, I think, why you recommend using the average on a whole week.

So the result of this PromQL query isn't actually the minimum number of queriers necessary but the scale factor that we need to apply to the **current** number of queriers in the cluster when we evaluated the inflight metric.

**tl;dr** If I understood your methodology correctly, you still need to add `* nb_querier` for this query to give the minimal number of querier required for the cluster to instantly handle a query in 75% of the time.
**Or** you can remove the sentence "running **_at least one_** queriers" for "running only one querier" as it means `nb_querier` would be 1, but it's not practical as most users will already have a running cluster with a specific number of querier to maintain the cluster in an equilibrium state.

Please, let met know if I am wrong in my understanding of your approach and thanks a lot for your work, it saved me a bunch of time ! :rocket:

Reply:
> [...] you want that 75% of the time, the average number of request in the system (L in queuing theory) is equal to 75% the number of workers per querier

Not quite. Using the 75 percentile (75% of the time) is to avoid taking short-lasting spikes into account.

> If I understood your methodology correctly, you still need to add * nb_querier for this query

I don't think so. We divide by the number of workers we aim to use so we have extra room for spikes.
E.g. if the avg is 10 and we aim to use 2 workers out of the 4 workers available per querier, this query would return 5 (the number of queriers). Because we'd be using 50% of the workers (2 out of 4), we have room for twice the avg workload.

Reply:
> @Raskyld Thanks for your comment! Sorry for the delay, I took some vacations.

No problem! Hope you enjoyed it :)

> Not quite. Using the 75 percentile (75% of the time) is to avoid taking short-lasting spikes into account.

Gotcha!

> I don't think so. We divide by the number of workers we aim to use so we have extra room for spikes. E.g. if the avg is 10 and we aim to use 2 workers out of the 4 workers available per querier, this query would return 5 (the number of queriers). Because we'd be using 50% of the workers (2 out of 4), we have room for twice the avg workload.

After two weeks and some sleep I realized my misunderstanding come from the metrics used.
Just for the record, and to be sure I understand correctly :

`cortex_query_scheduler_inflight_requests` observes (every `250ms`) the number of queries being processed by a querier worker or queued.
So you want the minimal number of worker = the average number of inflight requests 75% of the time ; then you divide the average by the number of active worker you want (e.g. 75% of all the workers) to know how much querier you need.

Am I correct? :pray: 

I have another concern : It is possible to replicate the `query-scheduler` as long as there is less or as much as instance of scheduler than instance of querier worker (i.e. `-querier.max-concurrent`) ; this means, you could have multiple series of this metric right? If so, shouldn't you sum the averages before dividing?

Reply:
> [...] the average number of inflight requests 75% of the time ; then you divide the average by the number of active worker you want (e.g. 75% of all the workers) to know how much querier you need.
> Am I correct? 

Yes! 

> you could have multiple series of this metric right? If so, shouldn't you sum the averages before dividing?

The `avg` is an aggregation function that will average all the series. To have different averages for each series you would do `avg by (instance)`.

There is an example [in the docs](https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators) that shows this:
> If the metric http_requests_total had time series that fan out by application, instance, and group labels, we could calculate the total number of seen HTTP requests per application and group over all instances via:
> `sum without (instance) (http_requests_total)`
> Which is equivalent to:
> `sum by (application, group) (http_requests_total)`
> If we are just interested in the total of HTTP requests we have seen in all applications, we could simply write:
> `sum(http_requests_total)`

Reply:
Guess it's my turn to say TIL!
`avg_over_time` uses an aggregation operator but does **not** aggregate while `avg` alone does!

Reply:
Thanks for all those explanations :ok_hand: